subsys/power/device_pm.c

```diff
@@ -0,0 +1,194 @@
+		}
+		break;
+	case DEVICE_PM_FSM_STATE_SUSPENDING:
+	case DEVICE_PM_FSM_STATE_RESUMING:
+		/* Do nothing: We are waiting for device_pm_callback() */
+		break;
+	default:
+		LOG_ERR("Invalid FSM state!!\n");
+	}
+
+	if (ret < 0) {
+		LOG_ERR("Set Power state error(%d)\n", ret);
+	}
+
+	return;
+
+fsm_out:
+	k_poll_signal_raise(&dev->config->pm->signal, pm_state);
+
+}
+
+static int device_pm_request(struct device *dev,
+			     u32_t target_state, u32_t pm_flags)
+{
+	int ret = 0;
+
+	__ASSERT((target_state == DEVICE_PM_ACTIVE_STATE) ||
+			(target_state == DEVICE_PM_SUSPEND_STATE),
+			"Invalid device PM state requested");
+
+	if (target_state == DEVICE_PM_ACTIVE_STATE) {
+		if (atomic_inc(&dev->config->pm->usage) < 0) {
+			goto req_out;
+		}
+	} else {
+		if (atomic_dec(&dev->config->pm->usage) > 1) {
+			goto req_out;
+		}
+	}
+
+	k_sem_take(&dev->config->pm->lock, K_FOREVER);
+	if (!dev->config->pm->enable) {
+		if (pm_flags & DEVICE_PM_ASYNC) {
+			/* Notify the driver through signal */
+			k_poll_signal_raise(&dev->config->pm->signal,
+								target_state);
+		}
+		k_sem_give(&dev->config->pm->lock);
+		goto req_out;
+	}
+	k_sem_give(&dev->config->pm->lock);
+
+	k_work_submit(&dev->config->pm->work);
+
+	/* Incase of Sync request wait for completion event */
+	if (!(pm_flags & DEVICE_PM_ASYNC)) {
+		u32_t signaled;
+		int result;
+
+		ret = k_poll(&dev->config->pm->event, 1, K_FOREVER);
+		if (ret < 0 || dev->config->pm->event.state
+					!= K_POLL_STATE_SIGNALED) {
+			ret = -EIO;
+		} else {
+			k_poll_signal_check(&dev->config->pm->signal,
+						&signaled, &result);
+			if (result != target_state) {
+				ret = -EIO;
+			}
+		}
+	}
+
+req_out:
+	return ret;
+}
+
+int device_pm_get(struct device *dev)
+{
+	return device_pm_request(dev,
+			DEVICE_PM_ACTIVE_STATE, DEVICE_PM_ASYNC);
+}
+
+int device_pm_get_sync(struct device *dev)
+{
+	return device_pm_request(dev, DEVICE_PM_ACTIVE_STATE, 0);
+}
+
+int device_pm_put(struct device *dev)
+{
+	return device_pm_request(dev,
+			DEVICE_PM_SUSPEND_STATE, DEVICE_PM_ASYNC);
+}
+
+int device_pm_put_sync(struct device *dev)
+{
+	return device_pm_request(dev, DEVICE_PM_SUSPEND_STATE, 0);
+}
+
+void device_pm_enable(struct device *dev)
+{
+	k_sem_take(&dev->config->pm->lock, K_FOREVER);
+	dev->config->pm->enable = true;
+
+	/* During the driver init, device can set the
+	 * PM state accordingly. For later cases we need
+	 * to check the usage and set the device PM state.
+	 */
+	if (!dev->config->pm->dev) {
+		dev->config->pm->dev = dev;
+		atomic_set(&dev->config->pm->fsm_state,
+					DEVICE_PM_FSM_STATE_SUSPENDED);
+		k_work_init(&dev->config->pm->work, pm_work_handler);
+	} else {
+		k_work_submit(&dev->config->pm->work);
+	}
+	k_sem_give(&dev->config->pm->lock);
+}
+
+void device_pm_disable(struct device *dev)
+{
+	k_sem_take(&dev->config->pm->lock, K_FOREVER);
+	/* Bring up the device before disabling the Idle PM */
+	k_work_submit(&dev->config->pm->work);
+	dev->config->pm->enable = false;
```

Comment:
I suggest moving `dev->config->pm->enable = false;` before `k_work_submit()`. If work queue thread has higher priority than caller thread, the work will be executed before enable flag change and the FSM will not update device state.

Reply:
wil fix it.